File: algorithm_library/python/math/geometry.py
import math


# (x1,y1) - (x2,y2) の正規化ベクトルを返す
def normalize_vector(x1, y1, x2, y2):
    dx = x1 - x2
    dy = y1 - y2
    if dx == 0:
        dx = 0
        dy = 1
    elif dy == 0:
        dx = 1
        dy = 0
    else:
        if dx < 0:
            dx *= -1
            dy *= -1
        g = math.gcd(abs(dx), abs(dy))
        dx //= g
        dy //= g
    return dx, dy


# (x0,y0) - (x1,y1) を通る直線の上に (px,py) があるか
def on_line(x0, y0, x1, y1, px, py) -> bool:
    return (y1 - y0) * (px - x0) == (x1 - x0) * (py - y0)


# 2点 (x0, y0), (x1, y1) を通る直線 ax + by + c = 0 の (a, b, c) を返す
def get_line_abc(x0, y0, x1, y1):
    a = y1 - y0
    b = x0 - x1
    c = -(a * x0 + b * y0)
    return a, b, c


def pos_after_t(sx, sy, gx, gy, t, v, dist=None):
    """
    始点(sx, sy) → 終点(gx, gy) を速さ v で動くとき、
    t 秒後の位置を返す（到達したらゴールにとどまる）。
    dist = 始点と終点の距離（前計算しといた方が速い）
    """
    # (dx,dy) = 方向ベクトル
    dx = gx - sx
    dy = gy - sy
    if dist is None:
        dist = math.sqrt(dx * dx + dy * dy)

    # 始点と終点が同じ
    if dist == 0 or v <= 0:
        return sx, sy

    # 進んだ距離
    move = v * t

    # すでに（またはちょうど）到達している
    if move >= dist:
        return gx, gy

    # 全体の何割進んだか
    u = move / dist

    # 方向ベクトルに進んだ距離をかける
    x = sx + dx * u
    y = sy + dy * u
    return x, y

File: algorithm_library/cpp-from-py/math/geometry.cpp
#include <bits/stdc++.h>
using namespace std;

pair<long long, long long> normalize_vector(long long x1, long long y1, long long x2, long long y2) {
    long long dx = x1 - x2, dy = y1 - y2;
    if (dx == 0) {
        dx = 0;
        dy = 1;
    } else if (dy == 0) {
        dx = 1;
        dy = 0;
    } else {
        if (dx < 0) dx *= -1, dy *= -1;
        long long g = gcd(abs(dx), abs(dy));
        dx /= g;
        dy /= g;
    }
    return {dx, dy};
}

bool on_line(long long x0, long long y0, long long x1, long long y1, long long px, long long py) {
    return (y1 - y0) * (px - x0) == (x1 - x0) * (py - y0);
}

tuple<long long, long long, long long> get_line_abc(long long x0, long long y0, long long x1, long long y1) {
    long long a = y1 - y0;
    long long b = x0 - x1;
    long long c = -(a * x0 + b * y0);
    return {a, b, c};
}

pair<double, double> pos_after_t(double sx, double sy, double gx, double gy, double t, double v, optional<double> dist = nullopt) {
    double dx = gx - sx, dy = gy - sy;
    double d = dist.value_or(sqrt(dx * dx + dy * dy));
    if (d == 0 || v <= 0) return {sx, sy};
    double move = v * t;
    if (move >= d) return {gx, gy};
    double u = move / d;
    return {sx + dx * u, sy + dy * u};
}
